### DISCUSSION TITLE
Status of the project

### DISCUSSION THREAD

**Discussion by ArchangeGabriel** [OpenSMTPD/OpenSMTPD#1152] [General]
Hi there,

I’ve seen that BSD has moved to 7.0, while portable is still at 6.8 (or even 6.7 according to https://www.opensmtpd.org/).

Following @poolpOrg [april report](https://poolp.org/posts/2021-04-30/april-2021-opensmtpd-plakar-ipcmsg-privsep-and-a-small-hypnosis-talk/) (and all the good things in [december](https://poolp.org/posts/2020-12-24/december-2020-opensmtpd-6.8.0p1-released-fixed-several-bugs-proposed-several-diffs-book-is-on-github/), [january](https://poolp.org/posts/2021-01-31/january-2021-opensmtpd-libtls-conversion-and-unix-domain-sockets-support-noosmtpd/) and [february](https://poolp.org/posts/2021-02-28/february-2021-noosmtpd-libtls-conversion-ciphers-curves-and-protocols/) ones) I had expected to see a call for testing the 6.9 portable with libtls, compat or not. Especially since I’m packaging both opensmtpd, libressl and libretls for Arch Linux, I’d be happy to help testing and ensure everything works OK in all expected cases (w.r.t. tls layer).

Now I understand that @poolpOrg might be less and less involved in OpenSMTPD for several reasons, and I’m still unsure what exact role has @ericfaurot when it comes to the portable version. That’s why I decided to open this discussion, to see whether the portable version is still something or if we would better consider it stalled. Once again I understand this is a very low bus factor project (and sadly I can’t contribute to it —I still hope to learn programming at some point, but won’t happen anytime soon…), and if you cannot handle it anymore that’s OK (I’m myself always struggling with my own duties), but I would just want to know it in that case. ;)

In any case I hope you’re well, that’s what matters the most! :)

**Comment by poolpOrg** [comment]
Hi,

I think this is a discussion worth having, so I'll just clarify a few things upfront.

First of all, I have distanced myself as the main developer for OpenSMTPD and from OpenBSD because I needed to reduce my stress level which was going off the roof. I told the OpenBSD/OpenSMTPD hackers that I'd still be around to help if needed and that I'd still submit and review diffs (which I did on multiple occasions) and we do have discussions on tech@ and also behind the scene about the future of OpenSMTPD. I'm still involved, just not as much as before... because that was unsustainable.

Which brings me to the point that's of interest for this discussion, and for explanations on why there was no 6.9 call for testing:

OpenBSD developers (and OpenSMTPD developers by extension) do not care much for other systems, not in the sense that they're not interested about getting code to work on them, but in the sense that when they write code, they only target OpenBSD. Upstream can accommodate small things to make it easier for -portable when requested, but when someone commits upstream, it is for OpenBSD's version of OpenSMTPD, assuming the OpenBSD base system and base dependencies (ie: the latest version of LibreSSL, or the latest changes in system headers).

There used to be a developer that bootstrapped portable and had an interest in it, but he left a few years ago, and with OpenSMTPD gaining in popularity, people started using -portable more and more which caused me to receive more and more mails, leading me to pick over his work and to start working both on leading OpenSMTPD _and_ dealing with -portable for all targets... a -portable version which I don't even use because I only run OpenBSD servers, this was work done 100% for the community and 0% for me, unlike work I do on filters and other features.

When OpenSMTPD 6.9 came out, it switched from OpenSSL API to libtls API. I had already done the work for -portable to abstract this change for systems that didn't link with LibreSSL (I did a libtls wrapper above OpenSSL) but for systems that had LibreSSL, it required the latest version of it... which was not yet available on any target. I asked on IRC but it seemed that people were waiting for OpenSMTPD to be released before packaging it, but I could not build OpenSMTPD because the package was not available, chicken and egg. There was not much visible difference between 6.8 and 6.9 so after we were already half-way through a release cycle, I decided to skip this release rather than throw myself a lot of work three months before doing it again for 7.0.

This OpenSSL -> libtls switch was a bit of a big flag day so it is not unexpected that things would go off track for a release or so, it is not representative of the amount of work generally required in -portable, but now with OpenSMTPD 7.0 out, I brought all commits from upstream (https://github.com/OpenSMTPD/OpenSMTPD/commits/master) and I'm still all by myself to maintain a portable version that I don't use and which requires dependencies that aren't available at this point. I am very happy to help with bringing diffs and resolving conflicts, but not so much on spending a lot of time resolving issues on systems that I don't use ... all by myself.

@ericfaurot and other OpenBSD developers do not really look at this repository or bug tracker, so when you see an issue fixed here, it is most likely because I fixed it myself and submitted upstream or because I proxied the issue and they fixed. As I'm spending less time on this project, if the community wants it to continue releasing as frequently as upstream (for years I released -portable on the exact same day as upstream), we need to have two people who have enough interest that they want to dedicate a few hours each month on this repository, otherwise it is just me and I'll continue dedicating time but will always be lagging.

The four hours I spent on merging 7.0 to -portable and fixing conflicts could have been reduced to minutes if people had helped me these last few months, but sadly no one did :-(

I don't know if the project is stalling but it is certainly lagging due to lack of {wo,}manpower

**Reply by ArchangeGabriel** [reply]
Thanks Gilles for this lengthy explanation!

I definitively understand why you skipped 6.9 and that was a good idea indeed (we still didn’t manage to bridge the new libera IRC to Matrix so I missed your messages, don’t hesitate to ping me —archange— on IRC when you are available so we can reinstate it). I didn’t know how you came into maintaining -portable nor that you actually didn’t use it anywhere (I thought like maybe for some clients or so, since anyway for your own use you’re probably still running some unreleased version of noOpenSMTPD), this is indeed far from an ideal situation.

I definitively missed that you committed changes recently (I currently have a bug on Firefox that makes both latest commits per path on the right and latest overall on top not appearing at all), thanks for doing that. :) Should we try compiling latest master against libressl/libretls (we have both at 3.4.1 on Arch) for release preparation?

As I said unfortunately I’m unable to contribute code and likely not able to merge from BSD to -portable (I don’t do C at all), but if there is a way I can help I’ll be happy to do so. Regarding helping with the code, maybe @DemiMarie that showed a lot of interest in the project some months ago and/or @sjbronner that much more recently expressed interest in it could both be interested in helping, I don’t know? At the very least they are the two people I could think of that showed some low level understanding of both RFC and OpenSMTPD code.

Finally, I want to say that indeed the project is not stalling thanks to you still dedicating time to it, but indeed lagging/slowly progressing albeit no one should expect better given that you, the only maintainer, don’t use the project anywhere so it’s already a great gift from you to actually spend some of your very little available time on it!

**Reply by DemiMarie** [reply]
@ArchangeGabriel I imagine @sjbronner would be a better choice for this; I have my hands full with my work for @QubesOS.  My interest mostly was in fixing various security concerns.

**Reply by ArchangeGabriel** [reply]
Alright, maybe @satan1st then, who PR’ed before and just did again. ;)

**Reply by ghost** [reply]
I'll check what I can do. Currently working on build for debian and freebsd.

**Comment by thenktor** [comment]
Some time has passed and I'm also interested in the state of the portable version. I've installed opensmtpd on FreeBSD and just noticed that is not "up to date" while looking for this error https://github.com/OpenSMTPD/OpenSMTPD/issues/1165

Other than that I'd like to thank all contributors for writing an MTA that finally has a config file that can be read and does not have a zillion of lines :+1:

**Reply by poolpOrg** [reply]
You came at the right time to ask this question.

The discussion was started a year and a half ago and there was absolutely no change in situation, which made me wonder if the project would not be discontinued outside of OpenBSD and if the repository should not be archived ...

... until earlier this month when @omar-polo started synchronising the repository with upstream and basically reviving the portable codebase which, as far as I know, is now up-to-date and needs testing.

**Reply by thenktor** [reply]
OK, at least it sounds more promising than I've feared :+1:

**Comment by ArchangeGabriel** [comment]
Just in case some other people would stumble here, thanks to amazing work from @omar-polo the project is currently revived and a new release has been made available. Congrats on that! <3
